### PR TITLE
Visualize cardinality of foreign key columns

### LINKED ISSUE #283: Visualize cardinality of foreign key columns
_I'm filing this issue on behalf of a user request on our [Slack](https://bit.ly/sdv-slack-invite)._

### Problem Description
Currently, users are able to plot the data in statistical columns such as numerical, categorical, etc. (`utils.get_relationship_plot`) only supports columns that are numerical, categorical, boolean or datetime.

It would be nice to support a visualization for the foreign key/primary key relationship -- when it comes to the cardinality.

### Expected behavior
Create a new visualization `utils.get_cardinality_plot`. This should plot the cardinality (# of children) that each parent row has, colored by real vs. synthetic data.

**Parameters**:
- (required) `real_data`: A dictionary mapping each table name to a [pandas.DataFrame](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) containing the data. This dictionary corresponds to real data.
- (required) `synthetic_data`: A dictionary mapping each table name to a [pandas.DataFrame](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) containing the data. This dictionary corresponds to synthetic data.
- (required) `child_table_name`: The string name of the child table
- (required) `parent_table_name`: The string name of the parent table
- (required) `child_foreign_key`: The string name of the child's foreign key column (that links to the parent's primary key)
- (required) `metadata`: A dictionary of Multi Table Metadata

Output: A plotly.Figure object with a bar graph. The graph shows the # of children that each parent row has. The color represents real vs. synthetic data.

```python
from sdmetrics.reports import utils

fig = utils.get_cardinality_plot(
    real_data=real_tables,
    synthetic_data=synthetic_tables,
    parent_table_name='users',
    child_table_name='transactions',
    child_foreign_key='user_id',
    metadata=my_multi_table_metadata
)

fig.show()
```

### PR DESCRIPTION
**Pull request by R-Palazzo**
Resolve #283

### CODE REVIEW

**Review by fealho — CHANGES_REQUESTED** [review]
If I'm understanding things correctly, `get_cardinality` can be simplified a bit. Besides that it looks pretty good 👍

**Comment by R-Palazzo** [comment]
Thanks for the review @fealho! I addressed the comment in [5ed67c3](https://github.com/sdv-dev/SDMetrics/pull/384/commits/5ed67c3b9fde229f639062099fc701dc3717c5c4)

File: sdmetrics/reports/utils.py [hunk 1]
```diff
@@ -479,6 +479,124 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
...
+    fig.update_layout(
+        title=title,
+        xaxis_title='# of Children (per Parent)',
+        yaxis_title='Frequency',
+        plot_bgcolor=BACKGROUND_COLOR
+    )
+    return fig
+
+
+def get_cardinality_plot(real_data, synthetic_data, child_table_name, parent_table_name,
```
Comment (fealho) [inline sdmetrics/reports/utils.py:546]: Add `Return` to docstring of all three methods.

**Review by fealho — COMMENTED** [review]
Thanks for addressing, the code looks much more concise now! I tested locally and it seems good, I just wanted to ask you why are you doing the `'# parents'` stuff? It doesn't seem like it gets added to the final image.

**Comment by R-Palazzo** [comment]
> Thanks for addressing, the code looks much more concise now! I tested locally and it seems good, I just wanted to ask you why are you doing the `'# parents'` stuff? It doesn't seem like it gets added to the final image.

The `'# parents'` column corresponds to the y-axis in the graph, but we plot the density rather than a count.

File: sdmetrics/reports/utils.py (hunk 1 above)
Comment (R-Palazzo) [reply sdmetrics/reports/utils.py:552]: Yes done in [1de3440](https://github.com/sdv-dev/SDMetrics/pull/384/commits/1de3440d8b0a3f80de23c5d3b0ec909796e23242)

File: sdmetrics/reports/utils.py [hunk 2]
```diff
@@ -479,6 +479,133 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
...
+    fig = px.histogram(data,
+                       x='# children',
+                       y='# parents',
+                       color='data',
+                       barmode='group',
+                       color_discrete_sequence=[DATACEBO_DARK, DATACEBO_LIGHT],
+                       pattern_shape='data',
+                       pattern_shape_sequence=['', '/'],
+                       nbins=max(data['# children']) - min(data['# children']) + 1,
+                       histnorm='probability density')
```
Comment (amontanez24) [inline sdmetrics/reports/utils.py:531]: minor: can we start all the parameters a line below and not as tabbed over

File: sdmetrics/reports/utils.py [hunk 3]
```diff
@@ -479,6 +479,133 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
         return make_mixed_column_pair_plot(real_data, synthetic_data)
 
 
+def get_cardinality(parent_table, child_table, parent_primary_key, child_foreign_key):
```
Comment (amontanez24) [inline sdmetrics/reports/utils.py:482]: can we make this private for now?

File: sdmetrics/reports/utils.py [hunk 4]
```diff
@@ -479,6 +479,133 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
...
+    child_per_parent = child_counts.reindex(parent_table[parent_primary_key], fill_value=0)
+    child_per_parent = child_per_parent.reset_index()
+    child_per_parent.columns = [parent_primary_key, '# children']
+
+    cardinalities = child_per_parent.groupby('# children').size().reset_index(name='# parents')
+
+    return cardinalities.sort_values('# children')
+
+
+def generate_cardinality_plot(data, parent_primary_key, child_foreign_key):
```
Comment (amontanez24) [inline sdmetrics/reports/utils.py:508]: can we make this private too?

File: sdmetrics/reports/utils.py [hunk 5]
```diff
@@ -479,6 +479,133 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
...
+        child_foreign_key (string):
+            The name of the foreign key column in the child table.
+        metadata (dict):
+            The metadata.
+
+    Returns:
+        plotly.graph_objects._figure.Figure
+    """
+    parent_primary_key = None
+    for relation_dict in metadata.get('relationships', []):
```
Comment (amontanez24) [inline sdmetrics/reports/utils.py:574]: minor: I don't think we need the dict suffix

File: sdmetrics/reports/utils.py [hunk 6]
```diff
@@ -479,6 +479,133 @@ def get_column_pair_plot(real_data, synthetic_data, column_names, metadata):
...
+        child_match = relation_dict['child_table_name'] == child_table_name
+        foreign_key_match = relation_dict['child_foreign_key'] == child_foreign_key
+        if child_match and parent_match and foreign_key_match:
+            parent_primary_key = relation_dict['parent_primary_key']
+
+    if parent_primary_key is None:
+        raise ValueError(
+            f"Relationship between child table '{child_table_name}' and parent table "
+            f"'{parent_table_name}' for the foreign key '{child_foreign_key}' "
+            'not found in the metadata. Please update the metadata.'
```
Comment (amontanez24) [inline sdmetrics/reports/utils.py:585]: minor: can we reword this to say "No relationship found between... in the metadata. Please update the metadata"

**Review by amontanez24 — CHANGES_REQUESTED** [review]

File: sdmetrics/reports/utils.py (hunk 6 above)
Comment (R-Palazzo) [reply sdmetrics/reports/utils.py]: Yes, done in [76fcf86](https://github.com/sdv-dev/SDMetrics/pull/384/commits/76fcf862286df85acac9ee2977216ac5c72f5a30)

File: sdmetrics/reports/utils.py (hunk 5 above)
Comment (R-Palazzo) [reply sdmetrics/reports/utils.py]: Yes, done in [76fcf86](https://github.com/sdv-dev/SDMetrics/pull/384/commits/76fcf862286df85acac9ee2977216ac5c72f5a30)

File: sdmetrics/reports/utils.py (hunk 3 above)
Comment (R-Palazzo) [reply sdmetrics/reports/utils.py]: Yes I was thinking of this first, but then I saw that we don't do it in this file. Done in [76fcf86](https://github.com/sdv-dev/SDMetrics/pull/384/commits/76fcf862286df85acac9ee2977216ac5c72f5a30)

File: sdmetrics/reports/utils.py (hunk 2 above)
Comment (R-Palazzo) [reply sdmetrics/reports/utils.py]: Yes, done in [76fcf86](https://github.com/sdv-dev/SDMetrics/pull/384/commits/76fcf862286df85acac9ee2977216ac5c72f5a30)

File: sdmetrics/reports/utils.py (hunk 4 above)
Comment (R-Palazzo) [reply sdmetrics/reports/utils.py]: Yes, done in [76fcf86](https://github.com/sdv-dev/SDMetrics/pull/384/commits/76fcf862286df85acac9ee2977216ac5c72f5a30)

**Review by amontanez24 — APPROVED** [review]
:shipit:
